lib/ActivityStreams.js

```diff
@@ -132,17 +142,19 @@ ActivityStreams.prototype = {
   },
 
   /**
-   * Broadcast places changes to pages
+   * Handles changes to places
    */
   _handlePlacesChanges(eventName, data) {
-    this.broadcast(am.actions.Response("RECEIVE_PLACES_CHANGES", data));
-  },
-
-  /**
-   * Broadcast bookmark changes to pages
-   */
-  _handleBookmarkChanges(eventName, data) {
-    this.broadcast(am.actions.Response("RECEIVE_BOOKMARKS_CHANGES", data));
+    /* note: this will execute for each of the 3 notifications that occur
+     * when adding a visit: frecency:-1, frecency: real frecency, title */
+    this._invalidatePlacesCache();
+    this._asyncBuildPlacesCache();
+
```

Comment:
What would happen if a user opens a page and does "Command Click" on 10 links from this page?
That will open 10 tabs, and will generate 10 history events, which in turn may cause cascading invalidate and cache rebuilds calls.  Not sure how bad it is, but at the very list we should not clean the cache if we are rebuilding it.  Something like:
```
_invalidatePlacesCache() {
if (_building_cache_flag) { 
   // do not disturb
   return;
}
_building_cache_flag = true;
// build the cache
....
// done with cache
_building_cache_flag = false;
```
It would be good to unit test this use case to make sure we don't get in some crazy concurrency situation when cache is cleaned while being rebuilt.